src/node_crypto.cc

```diff
@@ -2563,6 +2564,841 @@ static bool IsSupportedAuthenticatedMode(int mode) {
+                                        const PublicKeyEncodingConfig& config) {
+  BIOPointer bio(BIO_new(BIO_s_mem()));
+  CHECK(bio);
+
+  bool err;
+
+  if (config.type_.ToChecked() == kKeyEncodingPKCS1) {
+    // PKCS#1 is only valid for RSA keys.
+    CHECK_EQ(EVP_PKEY_id(pkey), EVP_PKEY_RSA);
+    RSAPointer rsa(EVP_PKEY_get1_RSA(pkey));
+    if (config.format_ == kKeyFormatPEM) {
+      // Encode PKCS#1 as PEM.
+      err = PEM_write_bio_RSAPublicKey(bio.get(), rsa.get()) != 1;
+    } else {
+      // Encode PKCS#1 as DER.
+      CHECK_EQ(config.format_, kKeyFormatDER);
+      err = i2d_RSAPublicKey_bio(bio.get(), rsa.get()) != 1;
+    }
+  } else {
+    CHECK_EQ(config.type_.ToChecked(), kKeyEncodingSPKI);
+    if (config.format_ == kKeyFormatPEM) {
+      // Encode SPKI as PEM.
+      err = PEM_write_bio_PUBKEY(bio.get(), pkey) != 1;
+    } else {
+      // Encode SPKI as DER.
+      CHECK_EQ(config.format_, kKeyFormatDER);
+      err = i2d_PUBKEY_bio(bio.get(), pkey) != 1;
+    }
+  }
+
+  if (err) {
+    ThrowCryptoError(env, ERR_get_error(), "Failed to encode public key");
+    return MaybeLocal<Value>();
+  }
+  return BIOToStringOrBuffer(env, bio.get(), config.format_);
+}
+
+static EVPKeyPointer ParsePrivateKey(const PrivateKeyEncodingConfig& config,
+                                     const char* key,
+                                     size_t key_len) {
+  EVPKeyPointer pkey;
+
+  if (config.format_ == kKeyFormatPEM) {
+    BIOPointer bio(BIO_new_mem_buf(key, key_len));
+    CHECK(bio);
+
+    char* pass = const_cast<char*>(config.passphrase_.get());
+    pkey.reset(PEM_read_bio_PrivateKey(bio.get(),
+                                       nullptr,
+                                       PasswordCallback,
+                                       pass));
+  } else {
+    CHECK_EQ(config.format_, kKeyFormatDER);
+
+    if (config.type_.ToChecked() == kKeyEncodingPKCS1) {
+      const unsigned char* p = reinterpret_cast<const unsigned char*>(key);
+      pkey.reset(d2i_PrivateKey(EVP_PKEY_RSA, nullptr, &p, key_len));
+    } else if (config.type_.ToChecked() == kKeyEncodingPKCS8) {
+      BIOPointer bio(BIO_new_mem_buf(key, key_len));
+      CHECK(bio);
+      char* pass = const_cast<char*>(config.passphrase_.get());
+      pkey.reset(d2i_PKCS8PrivateKey_bio(bio.get(),
+                                         nullptr,
+                                         PasswordCallback,
+                                         pass));
+    } else {
+      CHECK_EQ(config.type_.ToChecked(), kKeyEncodingSEC1);
+      const unsigned char* p = reinterpret_cast<const unsigned char*>(key);
+      pkey.reset(d2i_PrivateKey(EVP_PKEY_EC, nullptr, &p, key_len));
+    }
+  }
+
+  // OpenSSL can fail to parse the key but still return a non-null pointer.
+  if (ERR_peek_error() != 0)
+    pkey.reset();
+
+  return pkey;
+}
+
+ByteSource::ByteSource() : data_(nullptr), allocated_data_(nullptr), size_(0) {}
```

Comment:
I’d have a slight preference to give these initial values in the class definition, so that we can just use the default constructor